src/js/testem-component.js

```diff
@@ -0,0 +1,531 @@
+        fluid.log("Triggering Testem shutdown...");
+        callback();
+    });
+    that.events.stopFixtures.fire();
+};
+
+/**
+ *
+ * Optionally instrument the source code under test.
+ *
+ * @param that - The component itself.
+ * @param eventCallback - The event callback to fire when our work is complete.
+ *
+ */
+gpii.testem.instrumentAsNeeded = function (that, eventCallback) {
+    if (that.options.instrumentSource) {
+        try {
+            // Create our instrumentation directory if it doesn't already exist.
+            mkdirp.sync(that.options.instrumentedSourceDir);
+            fluid.each(fluid.makeArray(that.options.sourceDirs), function (sourcePath) {
+                var resolvedSourcePath = path.resolve(that.options.testemOptions.cwd, sourcePath);
+                var parsedPath = path.parse(resolvedSourcePath);
+
+                var pathStats = fs.statSync(resolvedSourcePath);
+                var lastDirSegment = pathStats.isDirectory() ? parsedPath.base : path.baseName(parsedPath.dir);
+
+                var targetPath = path.resolve(that.options.instrumentedSourceDir, lastDirSegment);
+
+                // Instrument each directory to its own subdirectory using a command like:
+                // istanbul instrument --output /tmp/instrumentSource/src src
+                var commandSegments = ["istanbul instrument --output", targetPath, resolvedSourcePath, "--complete-copy"];
+                var command = commandSegments.join(" ");
+
+                exec(command, function (error) {
+                    if (error) {
+                        console.error("Error running instrumentation command:", error);
+                    }
+                });
+
+                // Add a "route" so that the instrumented code will seamlessly replace its uninstrumented counterpart.
+                that.generatedOptions.routes[path.join("/", lastDirSegment)] = path.join("instrumented", lastDirSegment);
+
+                return targetPath;
+            });
+        }
+        catch (error) {
+            console.error("Error instrumenting code:", error);
+        }
+    }
+
+    eventCallback();
+};
+
+/**
+ *
+ * Deliver our effective options to Testem.  We do this this way to avoid exposing options to Testem before they have
+ * been completely assembled.
+ *
+ * @param that - The component itself.
+ * @returns {Object} - The Testem options to use for this run.
+ *
+ */
+gpii.testem.getTestemOptions = function (that) {
+    return fluid.extend(that.options.testemOptions, that.generatedOptions);
+};
+
+/**
+ *
+ * Remove all Testem browser data from this run.
+ *
+ * @param path {String} - The path to the directory which contains Testem's browser data from this run.
+ * @returns {Promise} - A promise that will be resolved when cleanup is complete, or rejected if there is an error.
+ *
+ */
+gpii.testem.cleanupTestemContent = function (path) {
+    var testemRegexp = /testem-.+/;
+    var togo = fluid.promise();
+
+    try {
```

Comment:
The bodies of functions in the file seem to be written to a kind of "recurring pattern" in that they do something fallible, catch an exception and then convert it to a promise rejection. I am speculating on what it might mean for this kind of workflow to be factored out into a common utility. I can't see any clear way of doing this that offers any particular benefits, but thought that I would just drop this thought in here for us to ponder about it.